R/api.R

```diff
@@ -0,0 +1,136 @@
+##' Run a pkgbuilder server
+##'
+##' @title Run a pkgbuilder server
+##'
+##' @param versions A vector of 2-digit version numbers corresponding
+##'   to the versions of R to build packages for.  Workers will need
+##'   to be provided separately to build the packages.
+##'
+##' @param workdir A directory to work in. This must be shared between
+##'   the workers and the server, so do not use a temporary directory.
+##'
+##' @param port The port to run on. Your operating system will likely
+##'   restrict you from using very low port numbers, so use something
+##'   like 8080
+##'
+##' @return Never returns - runs a HTTP server as a side-effect
+##' @export
+pb_server <- function(versions, workdir, port) {
+  queue <- queue$new(versions, workdir)
+  api <- api_build(queue)
+  api$run(port = port)
+}
+
+
+api_build <- function(queue) {
+  api <- pkgapi::pkgapi$new()
+  api$handle(endpoint_root(queue))
+  api$handle(endpoint_submit(queue))
+  api$handle(endpoint_list(queue))
+  api$handle(endpoint_status(queue))
+  api$handle(endpoint_result(queue))
+  api
+}
+
+
+target_root <- function(queue) {
+  pkgbuilder <- scalar(as.character(utils::packageVersion("pkgbuilder")))
+  function() {
+    list(pkgbuilder = pkgbuilder,
+         versions = queue$versions)
+  }
+}
+
+
+endpoint_root <- function(queue) {
+  returning <- pkgapi::pkgapi_returning_json("Root.schema", schema_root())
+  pkgapi::pkgapi_endpoint$new(
+    "GET", "/", target_root(queue),
+    returning = returning)
+}
+
+
+target_list <- function(queue) {
+  function(version) {
+    queue$list(version)
+  }
+}
+
+
+endpoint_list <- function(queue) {
+  returning <- pkgapi::pkgapi_returning_json("List.schema", schema_root())
+  pkgapi::pkgapi_endpoint$new(
+    "GET", "/<version>/list", target_list(queue),
+    returning = returning)
+}
+
+
+target_submit <- function(queue) {
+  function(version, data) {
+    data <- from_json(data)
+    id <- queue$submit(version, data$ref, data$additional_dependencies)
+    scalar(id)
+  }
+}
+
+
+endpoint_submit <- function(queue) {
+  returning <- pkgapi::pkgapi_returning_json("Submit.schema", schema_root())
+  pkgapi::pkgapi_endpoint$new(
+    "POST", "/<version>/submit/ref", target_submit(queue),
```

Comment:
The idea here is that eventually this might support a direct upload endpoint where the user sends the package to build as a binary; this version builds off of a reference so I've made the endpoint `/submit/ref` rather than plain `/submit`